### DISCUSSION TITLE
token_stats.py - what does it do (and how)?

### DISCUSSION THREAD

**Discussion by krisstud** [facebookresearch/ParlAI#4998] [Q&A]
It seems that this script outputs the tokenized statistics of whole episodes, while data_stats gives statistics per turn. Is this correct?

Getting the statistics is horribly slow, and as far as i can tell does not use cuda. If my assumptions are correct, why is this so? Right now my best estimate is that i need more than 48 hours to generate statistics for my dataset using the CPU.

Changing the code to `self.opt['no_cuda'] = False` does not change this behavior.

**Comment by mojtaba-komeili** [comment]
I don't think this script uses any operation that can be accelerated via GPU. It mostly needs raw python string processing and dictionary operations for tokenizing the text string. So, using GPU will not help in speeding it up.

**Reply by krisstud** [reply]
Maybe my confusion stems from Huggingface's GPU acceleration of tokenization to vectors. If i understand you correctly, this script only converts the text strings into the equivalent indexes or values from the tokenizer vocab before calculating statistics?

**Reply by mojtaba-komeili** [reply]
I am not aware of Huggingface's GPU tokenization. A quick look online didn't surface anything either. Would you mind posting a link to their implementation of that?
As far as ParlAI goes, we have a simple tokenizer that uses Python string manipulation. Tokenization is a time-consuming process, but you only need to run it on each dataset (or part of it) once, and then you can reuse it. In fact, most of the times you may not need to have a new tokenizer and can use standard ones coming with models such as GPT2, which we have included in ParlAI.

**Reply by krisstud** [reply]
You're right. The only reference is placement of data in GPU memory and the torch format. 

The speed difference is night and day, though. Getting the stats using token_stats required 50 hours for my dataset. Tokenization of the same dataset (no statistics per sample, but that should be minor) using HF implementation took 25 sec. The example count (number of lines) is the same when processing using either approach.

Using token_stats:
`parlai token_stats --task custom_parlaiformatted_dataset_task --init-model 'zoo:bb3/bb3_3B/model' --fp16 True --log-every-n-secs 100`

Using HF implementation:
```
from transformers import GPT2TokenizerFast
import pandas as pd
from datasets import load_dataset
import timeit

tokenizer = GPT2TokenizerFast.from_pretrained("gpt2")

def preprocess_function(batch_sentences):
    return tokenizer(batch_sentences['text'], truncation=False)

# the delimiter is technically incorrect
df_import = pandas.read_csv(path, delimiter='\nepisode_done:True', names=['text'])
# quick cleanups
df_import = df_import.replace('^text:', '', regex=True)
df_import = df_import.replace('\s+labels:', ' ', regex=True)
df_import = df_import.replace('\t', '', regex=True)
dataset = Dataset.from_pandas(df_import)

starttime = timeit.default_timer()
dataset.map(preprocess_function, batched = True, batch_size = 16, load_from_cache_file=False)
print(timeit.default_timer() - starttime)
```

**Reply by klshuster** [reply]
I'm not really sure why your dataset took so long - when you run `parlai token_stats` on any other ParlAI dataset, does it also take that long?

I just ran your exact command with the `msc` dataset (236k utterances) and it took like 70 seconds. 

Also, specifying `--init-model` with this script won't work like you expect it to. To tokenize according to `bb3`, you'd want to specify `--dict-tokenizer gpt2`. This still takes roughly the same amount of time

**Reply by klshuster** [reply]
If you want to try using HF's "fast" tokenization, you can [change this line](https://github.com/facebookresearch/ParlAI/blob/feab5717d0fd5c198f9b68515670ac251feb6cf3/parlai/agents/hugging_face/dict.py#L165) to use `GPT2TokenizerFast` like you have in your script above

**Reply by krisstud** [reply]
Thank you for the correction. I couldn't quite understand why the --init-model param was there from the documentation, so i assumed it carried a reference to the correct tokenizer for that model. The result is the same using the corrected param. I can see that it's just one thread bouncing around the CPU at 100% load. Using the HF implementation i get ~25% load on all threads. This can't be the only explanation though.

My dataset is 351 670 utterances in the ParlAI format. It passes verify and i can start training with it. The MSC dataset is fast, as expected. SaferDialogues is too small to effectively test with. I started the thread so that others might benefit. I have my statistics, and i assume (hope) this does not affect tokenization during training.

**Reply by klshuster** [reply]
> the MSC dataset is fast, as expected

For `token_stats`? And yes agreed, there is something else going on here.

**Reply by krisstud** [reply]
> > the MSC dataset is fast, as expected
> 
> For `token_stats`? And yes agreed, there is something else going on here.

Yes. If it's any help, building the dictionary took seconds using:
```
!parlai build_dict --task custom_parlaiformatted_dataset_task --dict-file outpath
Building dictionary: 100%|██████████████████| 270k/270k [00:04<00:00, 62.0kex/s]
```
